posthog-react-native/src/posthog-rn.ts

```diff
@@ -87,31 +102,32 @@ export class PostHog extends PostHogCore {
       this.flush()
     })
 
-    if (this._persistence === 'file') {
+    if (this._persistence === 'async') {
       if (!storage) {
         console.warn(
           'PostHog was initialised without using PostHog.initAsync - this can lead to race condition issues.'
         )
       }
-
-      this._semiAsyncStorage =
-        storage || new SemiAsyncStorage(options?.customAsyncStorage || buildOptimisiticAsyncStorage())
+    } else {
+      this.setup()
+    }
+    if (this._persistence === 'memory') {
+      this._storage = new PostHogMemoryStorage()
+    } else {
+      this._storage =
+        storage || new PostHogSemiAsyncStorage(options?.customAsyncStorage || buildOptimisiticAsyncStorage())
     }
 
     // Ensure the async storage has been preloaded (this call is cached)
-
-    const setupAsync = async (): Promise<void> => {
-      if (!this._semiAsyncStorage?.isPreloaded) {
-        await this._semiAsyncStorage?.preloadAsync()
-      }
+    const setup = () => {
       this.setupBootstrap(options)
 
       // It is possible that the old library was used so we try to get the legacy distinctID
-      if (!this._semiAsyncStorage?.getItem(PostHogPersistedProperty.AnonymousId)) {
+      if (!this._storage?.getItem(PostHogPersistedProperty.AnonymousId)) {
         const legacyValues = await getLegacyValues()
         if (legacyValues?.distinctId) {
-          this._semiAsyncStorage?.setItem(PostHogPersistedProperty.DistinctId, legacyValues.distinctId)
-          this._semiAsyncStorage?.setItem(PostHogPersistedProperty.AnonymousId, legacyValues.anonymousId)
+          this._storage?.setItem(PostHogPersistedProperty.DistinctId, legacyValues.distinctId)
+          this._storage?.setItem(PostHogPersistedProperty.AnonymousId, legacyValues.anonymousId)
         }
       }
```

Comment:
This actually is the only blocker, no idea about it, maybe we can move this to the initAsync? or could we do a .then instead of await?

Reply:
The problem with `then` is that it might be executed a bit after and `getItem` may be called before the `then` gets the chance to persist the data and read outdated data.